### ISSUE TITLE
Exposing MLIR

### ISSUE THREAD

**Issue by dwincort** [plaidml/plaidml#1310]
Hi,

I would love to be able to examine the intermediate representations (various dialects of MLIR, I presume) that get produced during the PlaidML process, and I'm wondering if that's possible.  I can't seem to find documentation about it (sorry if that's my fault for not looking hard enough), and I'm hoping someone can point me in the right direction.

Thanks!

**Comment by flaub** [comment]
If you're using the HEAD of the `plaidml-v1` branch, then you can use a recently added feature to dump the IR after each pass is executed. This can be done using `PLAIDML_DUMP=/path/to/dir`. Additionally, many of us use `PLAIDML_VERBOSE=3` which will print the IR after each pass.

**Comment by dwincort** [comment]
Hi again,

Thanks for your super quick response last time, but I think I need more help than I first thought.  I failed to point out last time that I haven't used plaidml before and don't really know what I'm doing.

Following the "building from source" directions, I managed to build HEAD on `plaidml-v1`, and I got the python wheel built too (is it bad that I found the wheel at `bazel-bin/plaidml/wheel.tmp/tmp/dist/plaidml-0.0.0.dev0-py2.py3-none-manylinux1_x86_64.whl` instead of `.../wheel.pkg/dist/*.whl`?), but the keras wheel fails to build.  `plaidml-setup` worked fine, and a bunch of tests passed, so I figured I could just press on.  Next, I entered a virtualenv, used `pip install` to install the python wheel I just built and then `pip install`ed the keras stuff the normal way.  I'm able to run `python program.py` and get my program to run, but despite having
```python
os.environ["KERAS_BACKEND"] = "plaidml.keras.backend"
```
at the top of the file, I'm not convinced it's actually using plaidml.  There's lots of output about tensorflow, but nothing about plaidml.  (Also, adding `PLAIDML_VERBOSE=3` and/or `PLAIDML_DUMP=dump` before the command doesn't seem to have any effect.)

Any help would be much appreciated!

**Comment by brianretford** [comment]
Wee have tested this w/ Plaidml-v1, but we've also just changed the way device selection works (no longer uses plaidml-setup) and I'm not certain what else you'd need to do to make things work from PIPs (we'll get there soon).

You should be able to use bazel test //plaidbench:bin -- keras to get keras working w/ plaidml via the examples already in keras. We will be providing onnxruntime and XLA backends in the nearish future

**Comment by flaub** [comment]
To be clear, we still plan on using `plaidml-setup` :)

**Comment by flaub** [comment]
To that end, you might need to run:

```
bazelisk run //plaidml:setup
```

Then follow the prompts to select the device you want to run on. This part of the code is in flux, so it might not all work exactly right yet. But selecting the `llvm_cpu.0` device should work.

**Comment by flaub** [comment]
Hmm, so reading more closely your response, it looks like the keras wheel didn't build. Do you have any error messages you can share for that failure? That seems to be the root cause AFACIT.

**Comment by dwincort** [comment]
So, when I tried to run `bazelisk build //plaidml/keras:wheel` like it says in the instructions (https://plaidml.github.io/plaidml/docs/building.html#build-the-plaidml-keras-wheel), I got an error saying that the BUILD file couldn't be found.  Indeed, there is no `plaidml/keras` directory any more.

I literally _just_ thought as I was writing the above, "Huh, I wonder if it's somewhere else" and proceeded to locate the build info under `plaidml/bridge/keras` :facepalm:.  I feel super silly.  Anyway, it's building now.

Thanks again for all of your above help.  After this builds, I'll go back to trying your initial suggestion (that is, `PLAIDML_VERBOSE=3`) and utilize your other tips as necessary.
